apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,921 @@
-      currentPanelContainer = document.getElementById(
-        currentPanelInfo.container);
-    }
+/**
+ * Used in `waitForSlideAnimation` to help solving race conditions.
+ * @type {Defer}
+ */
+var previousAnimationDefer;
+
+/**
+ * Wait for the end of an animation on `panelElement`.
+ *
+ * @param {Element} panelElement The element to watch.
+ * @returns {Promise} Resolves when the animation ends.
+ */
+function waitForSlideAnimation(panelElement) {
+  if (previousAnimationDefer) {
+    previousAnimationDefer.reject(new Error('A new animation started'));
+  }
 
-    var promise = Promise.resolve();
+  var defer = Utils.Promise.defer();
+  previousAnimationDefer = defer;
 
-    // beforeLeave and beforeEnter can return a rejected promise to prevent
-    // sliding. An exception is never expected though.
-    if (currentPanelObject && currentPanelObject.beforeLeave) {
-      promise = promise.then(
-        currentPanelObject.beforeLeave.bind(currentPanelObject, transitionArgs)
-      );
-    }
+  panelElement.addEventListener('animationend', function onAnimationEnd() {
+    this.removeEventListener('animationend', onAnimationEnd);
+    defer.resolve();
+  });
+
+  setTimeout(defer.resolve, 500);
+  return defer.promise.then(() => previousAnimationDefer = null);
+}
+
+/**
+ * Layouts the new view so that the new view's beforeEnter can deal with a
+ * rendered view.
+ */
+function prepareSwitchPanel() {
+  debug('prepareSwitchPanel()');
+  var from = navigationTransition.states.from;
+  var to = navigationTransition.states.to;
+
+  var newView = to.panel;
+
+  var shouldPrepare = !from ||
+    (from.panel && newView &&
+     from.panel.behavior !== newView.behavior);
+
+  if (!shouldPrepare) {
+    return;
+  }
 
-    if (nextPanelObject.beforeEnter) {
-      promise = promise.then(
-        nextPanelObject.beforeEnter.bind(nextPanelObject, transitionArgs)
+  var newPanelElement = document.querySelector(`.panel-${newView.behavior}`);
+  if (!newPanelElement) {
+    throw new Error(
+      'Couldn\'t find the container element for view ' + newView.name
+    );
+  }
+
+  newPanelElement.classList.add('panel-prepare-activation');
+}
+
+/**
+ * Hides current panel, shows new panel.
+ * It will use an animation if we change panels in the same document.
+ *
+ * @returns {Promise} Resolves when the animation is over, if there is an
+ * animation. Otherwise returns a resolved promise.
+ */
+function switchPanel() {
+  debug('switchPanel()');
+  if (!navigationTransition) {
+    return Promise.reject(new Error(
+      'Trying to switch panel but no transition is started yet.'
+    ));
+  }
+
+  var from = navigationTransition.states.from;
+  var to = navigationTransition.states.to;
+
+  var oldView;
+  var oldPanelElement;
+
+  if (from) {
+    oldView = from.panel;
+    oldPanelElement = document.querySelector(`.panel-${oldView.behavior}`);
+    if (!oldPanelElement) {
+      throw new Error(
```

Comment:
__question:__ few lines above you use ```Promise.reject``` and here is ```throw```, do you have some smart rule when to choose one or another? :)